doc/crdt-details.md

```diff
@@ -0,0 +1,1086 @@
+    // the `false` below: don't invert undos since our first_candidate is based on the current undo set, not past
+    let mut deletes_from_union = self.deletes_from_union_before_index(first_candidate, false).into_owned();
+
+    for rev in &self.revs[first_candidate..] {
+        if let Edit { ref undo_group, ref inserts, ref deletes, .. } = rev.edit {
+            if groups.contains(undo_group) {
+                if !inserts.is_empty() {
+                    deletes_from_union = deletes_from_union.transform_union(inserts);
+                }
+            } else {
+                if !inserts.is_empty() {
+                    deletes_from_union = deletes_from_union.transform_expand(inserts);
+                }
+                if !deletes.is_empty() {
+                    deletes_from_union = deletes_from_union.union(deletes);
+                }
+            }
+        }
+    }
+
+    let deletes_bitxor = self.deletes_from_union.bitxor(&deletes_from_union);
+    let max_undo_so_far = self.revs.last().unwrap().max_undo_so_far;
+    (Revision {
+        rev_id: self.next_rev_id(),
+        max_undo_so_far,
+        edit: Undo { toggled_groups, deletes_bitxor }
+    }, deletes_from_union)
+}
+
+pub fn undo(&mut self, groups: BTreeSet<usize>) {
+    let (new_rev, new_deletes_from_union) = self.compute_undo(&groups);
+
+    let (new_text, new_tombstones) =
+        shuffle(&self.text, &self.tombstones, &self.deletes_from_union, &new_deletes_from_union);
+
+    self.text = new_text;
+    self.tombstones = new_tombstones;
+    self.deletes_from_union = new_deletes_from_union;
+    self.undone_groups = groups;
+    self.revs.push(new_rev);
+    self.rev_id_counter += 1;
+}
+```
+
+### Engine::gc
+
+This is a large function that is only used in the single-device case to limit the size of history and the size of `deletes_from_union` so that operations like insertion and undo are always fast even on large documents with long histories. It throws away information, so it is incorrect in the multi-device case, and so we have plans for optimizing all our representations and algorithms so that it is unnecessary and we can remove it. It's not particularly interesting, if you want to see how it works you can read the code.
+
+### Engine::merge
+
+This is the operation you were (maybe) waiting for! The CRDT merge operation that allows peer-to-peer syncing of edits in a conflict-free eventually-consistent way. It takes `self` and another instance of `Engine` and incorporates any changes which that `Engine` has into `self`.
+
+It does this by finding changes which the other `Engine` has but it doesn't and doing a whole bunch of transformation so that those edits can be appended directly on to the end of `self`'s list of `revs`. The append-only nature of merge preserves the ability for operations like `Engine::delta_rev_head` to work, and allows future optimizations of how things are persisted.
```

Comment:
i'd replace the second 'it' on this line with `self`. 

also missing an "s" on transformation.